casper/src/main/scala/coop/rchain/casper/merging/DagMerger.scala

```diff
@@ -3,59 +3,102 @@ package coop.rchain.casper.merging
 import coop.rchain.rholang.interpreter.RhoRuntime.RhoHistoryRepository
 import coop.rchain.rholang.interpreter.merging.RholangMergingLogic
-import coop.rchain.rholang.interpreter.merging.RholangMergingLogic.convertToReadNumber
+import coop.rchain.rholang.syntax._
 import coop.rchain.rspace.HotStoreTrieAction
 import coop.rchain.rspace.hashing.Blake2b256Hash
+import coop.rchain.rspace.merger.{ChannelChange, StateChange, StateChangeMerger}
 import coop.rchain.rspace.merger.EventLogMergingLogic.NumberChannelsDiff
-import coop.rchain.rspace.merger.{
-  ChannelChange,
-  EventLogMergingLogic,
-  StateChange,
-  StateChangeMerger
-}
-import coop.rchain.rspace.syntax._
+import coop.rchain.sdk.dag.merging.DagMergingLogic._
+import coop.rchain.sdk.dag.merging.data._
+import coop.rchain.sdk.dag.merging.DagMergingLogic
 import coop.rchain.shared.{Log, Stopwatch}
 import scodec.bits.ByteVector
+import coop.rchain.shared.syntax._
+import coop.rchain.rspace.syntax._
 
 object DagMerger {
 
-  def costOptimalRejectionAlg: DeployChainIndex => Long =
-    (r: DeployChainIndex) => r.deploysWithCost.map(_.cost).sum
-
-  /**
-    * Merges blocks finalized with the fringe
-    *
-    * @param fringeDiffBlocks all blocks included from previous fringe to new fringe
-    * @param previousFringeState base state for merging, previous fringe state hash
-    * @param index getter for block index
-    * @param historyRepository history repository instance
-    * @param rejectionCostF rejection cost calculation function
-    * @return a tuple (new fringe state hash, rejected deploys - signatures)
-    */
-  def mergeFringe[F[_]: Concurrent: Log](
-      fringeDiffBlocks: Set[BlockHash],
-      previousFringeState: Blake2b256Hash,
-      index: BlockHash => F[Vector[DeployChainIndex]],
+  private def resolveConflicts[F[_]: Concurrent: Log](
+      tips: Set[BlockHash],
+      finalFringe: Set[BlockHash],
+      finalState: Blake2b256Hash,
+      msgMap: Map[BlockHash, Message[BlockHash, Validator]],
+      acceptedFinally: Set[DeployChainIndex],
+      rejectedFinally: Set[DeployChainIndex],
       historyRepository: RhoHistoryRepository[F],
-      rejectionCostF: DeployChainIndex => Long
-  ): F[(Blake2b256Hash, Set[ByteString])] =
-    for {
-      actualSet <- fringeDiffBlocks.toList.traverse(index).map(_.flatten.toSet)
-      // TODO: support for late blocks (with justifications below previousFringeState)
-      lateSet = Set[DeployChainIndex]()
+      blockIndex: BlockHash => F[BlockIndex],
+      rejectionCost: DeployChainIndex => Long
+  ): F[Merge[Blake2b256Hash, DeployChainIndex]] = {
+    val cScope = conflictScope(tips, finalFringe, msgMap.getUnsafe(_: BlockHash).seen)
+    // lowest fringe required to to be able to resolve conflicts
+    val lFringe = lowestFringe(
+      cScope.map(m => msgMap.getUnsafe(m).fringe),
+      msgMap.getUnsafe(_: BlockHash).height
+    )
+    val fScope     = finalScope(finalFringe, lFringe, msgMap.getUnsafe(_: BlockHash).seen)
```

Comment:
Scope required for merging is conflict scope defined by fringe and latest messages + all messages down to lowest fringe across messages of conflict set. For this area block indices have to be created for node to be able to merge.

Reply:
We could split this functions in two. One to calculate the indices from the scope and this one can accept them.

In this way they can be more easily cached.